<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Using graph genomes for genome optimization · Mycelia.jl</title><link rel="canonical" href="https://cjprybol.github.io/Mycelia.jl/1.using-graph-genomes-for-genome-optimization/"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">Mycelia.jl</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Documentation</span><ul><li><a class="tocitem" href="../docstrings/">Docstrings</a></li></ul></li><li><span class="tocitem">Chapters</span><ul><li class="is-active"><a class="tocitem" href>Using graph genomes for genome optimization</a><ul class="internal"><li><a class="tocitem" href="#Introduction"><span>Introduction</span></a></li><li><a class="tocitem" href="#Experiment"><span>Experiment</span></a></li><li><a class="tocitem" href="#Results"><span>Results</span></a></li><li><a class="tocitem" href="#Conclusions"><span>Conclusions</span></a></li></ul></li><li><a class="tocitem" href="../2.iterative-graph-polishing/">Iterative Graph polishing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Chapters</a></li><li class="is-active"><a href>Using graph genomes for genome optimization</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Using graph genomes for genome optimization</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/cjprybol/Mycelia.jl/blob/master/docs/src/1.using-graph-genomes-for-genome-optimization.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Using-graph-genomes-for-genome-optimization"><a class="docs-heading-anchor" href="#Using-graph-genomes-for-genome-optimization">Using graph genomes for genome optimization</a><a id="Using-graph-genomes-for-genome-optimization-1"></a><a class="docs-heading-anchor-permalink" href="#Using-graph-genomes-for-genome-optimization" title="Permalink"></a></h1><h2 id="Introduction"><a class="docs-heading-anchor" href="#Introduction">Introduction</a><a id="Introduction-1"></a><a class="docs-heading-anchor-permalink" href="#Introduction" title="Permalink"></a></h2><p>Now that we have the ability to sequence as many genomes as the research community can afford to pay for, and that there are so many genomes publicly available online, we are now in a situation where we can learn what regions of genomes contribute to what kind of phenotypic properties by associating measured attributes with the co-occurrence of genetic features even for non-model and unculturable organisms. When building these models initially, there are no causal attributes that can necessarily be attributed because we are simply looking for correlative relationships of phenotypic traits appearing in the presence of or in the absence of specific genetic features. Given the ability to associate genetic features with phenotypic features that may be desirable or undesirable, we are given the opportunity to use graph pan genomes as vehicles for genetic optimization driven entirely by data. These non-causal relationships can be expanded upon over time with additional information that is derived from causal analysis conducted statistically and/or via direct experimentation.</p><p>Given a weighted graph structure that enables the attribution of specific expected properties to paths through the graph (genomes, transcriptomes, proteomes), there may be situations in which it may be desirable to modify and manipulate genomes to have them manifest these desirable characteristics in an actual organism. For example, it may be desired to encode features that are most consistent with outcomes such as health and longevity (humans), avirulence/lack of pathogenicity (pathogens/microbes/viruses), as well as having the smallest essential (minimal) genome (biomanufacturing).</p><p>Given a pangenomegraph with weights where the weights are proportional to an optimizeable measure such as correlation, likelihood, frequency, etc., the question becomes how do we find the optimal path through this genome graph such that the average score is maximized so that we can test whether the predictive value of the graph can manifest the desired outcomes if the genome was encoded in a living organism.</p><p>At the time of writing, the optimization algorithms that I am aware of only operate on the weights of the edges. For example, there are many optimization algorithms that find the shortest path between two nodes. Examples include A-star, djikstra, bellman-Ford, traveling salesman problem (note that traveling salesman problem assumes a cyclic optimization which is very helpful specifically when dealing with circular genomes such as those of phage bacteria or other prokaryotic organisms). Given that we don’t have a plethora of optimization algorithms that account for edge weights and node weights (aside from hidden Markov models which become impractical at very large graph sizes common in pangenomics), we will look at the node optimization and path optimization components of our goal individually. </p><p>We will first ask the question &quot;which nodes do we want to include?&quot;, which can be done by setting a threshold and only retaining nodes that meet that threshold, clustering (supervised and/or unsupervised), and likely many more methods. Given the set of nodes of interest, we can induce a subgraph containing just those nodes and then find the optimal or a sufficiently optimal path through the subset of remaining nodes that maximizes the overall optimization that we are trying to achieve.</p><p>For the problem of trying to find the minimally sized genome, our node optimization is relative frequency of genes contained in all known entities of a given species or other phylogenetic level of classification (threshold = must be contained in 100% of observed entities). In this same problem, the path optimization is the shortest viable path of observed possible paths that includes all of these nodes. Further possible optimization can be done if we are willing to consider connections that have not yet been observed (e.g. gene fusions and assumptions that non-coding elements between genes are themselves non-essential).</p><p>To create the structure necessary for this optimization, we will first build a pan genome graph that includes the DNA kmers, gene annotations + amino acid kmers, and the pathways that they encode. We can then look at the relative feature space in all three of these levels. The highest and most abstract level is that of the metabolic pathway (Note, in an ideal world all metabolic pathways would be fully resolved and validated and in such a world this would be very non-abstract, however in practice we are unable to unambiguously resolve and verify all metabolic pathways in all organisms and therefore I will consider it in the abstract sense for the purposes of this explanation). </p><p>[figure here showing pathways and relative frequency of conservation among a group of entities]</p><p>At a functional protein level, the goal is not necessarily to consider all possible proteins as unique variants but rather to cluster the proteins into functionally equivalent clades/families, ask which clades/families maximize our intended value (for minimal genome size, frequency in all entitities considered = 100%), and then after the clades have been selected, choose the optimal known variant of that clade that can then be selected in the AA space and further constrained into the DNA space based on codon frequency profiles. </p><p>[insert figure here showing protein families with relative frequencies and amino acid variability that then translates down into DNA space and alternate codon options that can be optimized again]</p><p>At the lowest level, we have the DNA itself where our goal is simply to try and maximize the optimization value of choice through all possible DNA pads in the graph until we have either created a completely circular circular rise full genome or we have reached the end of the chromosome all telomeres.</p><p>[insert figure here showing the traversal of a graph that either loops into a complete circle or terminates at terminal telomeres]</p><p>In all cases the final minimal form must go through a DNA genome (with the exception of RNA-based viruses, which would follow a conceptually equivalent format but would build on a foundation of an RNA pangenome/pantranscriptome). The key advantage of constraining in the pathway and protein layers of abstraction is the dimensionality reduction provided. A genome may contain billions of nucleic acid bases, but is likely to only contain up to thousands of pathways and thousands of genes. These constraints will enable more efficient optimization of incredibly large and complex pangenomes such as those of complex microbiomes (e.g. soil) and large eukaryotic genomes (e.g. humans and many crops).</p><p>Given a set of constrained points, which may or may not be equivalent to the set of all possible points depending on what kind of optimization steps were carried out before, we then try and find the optimal path between points in a stepwise (random or greedy) order and ask first &quot;is there a path between this next node and any of the current nodes in the graph?&quot; and if so, return that path and add this node to this graph. If not, add this node to the set of nodes in the graph and await connections to the remaining nodes yet to be evaluated.</p><p>[insert a figure showing this workflow where we have a full graph, a reduced set of nodes that meet the thresholding criteria, and then iterative making the connections between them]</p><p>(personal note, maybe could also use the Bellman-Ford on a given induced subgraph and just iteratively pick out the shortest connections?)</p><p>For the genome optimization problem, different sections of the graph will need to be sequestered away and isolated into chromosomes as we reach terminating steps. I think there are several possible terminating steps worth considering. One possible consideration is that we create a fully circularized chromosome. In the event of this happens, this should be an end condition for that chromosomes. When considering linear chromosomes, if we hit a terminal node in the graph in two locations, equivalent to reaching the telomeres biologically, we should sequester this node collection/chromosome off and remove considering additions to this chromosome for all future inclusions to our optimized graph.</p><p>When trying to decide if you’re multichromosome all genome is optimally complete I think there should be a sigmoidal binary supervised cluster that or something simpler such as a absolute threshold in whereby either we were able to segregate into two clear populations of optimal and non-optimal or we are able to fit a diminishing returns Vmax V not midpoint where we feel good that we have obtained half of or at least half of the core density of the system and we should let people modify this parameter to their suiting however by default we have it set to the larger of the 2K means clusters.</p><h2 id="Experiment"><a class="docs-heading-anchor" href="#Experiment">Experiment</a><a id="Experiment-1"></a><a class="docs-heading-anchor-permalink" href="#Experiment" title="Permalink"></a></h2><p>Create a compressed reference graph</p><p>theoretical: 3 pathways, two of which are important 3 proteins per pathway 1 sequence per 3 proteins 3 random sequences of length 100 - 1000 assign random importance weights to pathways</p><p>applied, known answer: minimal mycobacterium minimal synthetic yeast chromosome</p><p>applied, unknown answer: phage pathways phage vogs the largest genus of bacteriophage (EC tequatrovirus)</p><p>initial pass, work at the pathway level then start again, working at only the protein level then start again, working only at the dna level</p><h2 id="Results"><a class="docs-heading-anchor" href="#Results">Results</a><a id="Results-1"></a><a class="docs-heading-anchor-permalink" href="#Results" title="Permalink"></a></h2><p>Visualize the graph before and after simplification</p><h2 id="Conclusions"><a class="docs-heading-anchor" href="#Conclusions">Conclusions</a><a id="Conclusions-1"></a><a class="docs-heading-anchor-permalink" href="#Conclusions" title="Permalink"></a></h2><p>Assuming a direct, linear relationship between # of genes and cost on potentially wasteful metabolic expenditure, then every gene removal can increase the relative production of remaining metabolites.</p><p>While regulatory feedback loops make a direct relationship unlikely, if we operate under this train of thought, then creating minimal genomes should make organisms more efficient and plyable/flexible for bioengineering and minimize the likelihood of the production of any undesirable phenotypes and metabolic outputs if used as biofactories of pharmaceuticals.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../docstrings/">« Docstrings</a><a class="docs-footer-nextpage" href="../2.iterative-graph-polishing/">Iterative Graph polishing »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Wednesday 15 December 2021 20:16">Wednesday 15 December 2021</span>. Using Julia version 1.6.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
