<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Home · Mycelia.jl</title><link rel="canonical" href="https://cjprybol.github.io/Mycelia.jl/"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.0/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">Mycelia.jl</span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Home</a><ul class="internal"><li><a class="tocitem" href="#Install"><span>Install</span></a></li><li><a class="tocitem" href="#Related-Software:"><span>Related Software:</span></a></li><li><a class="tocitem" href="#Questions:"><span>Questions:</span></a></li><li><a class="tocitem" href="#Behaviors"><span>Behaviors</span></a></li><li><a class="tocitem" href="#Schema-for-a-pan-meta-genome-graph-with-annotations-(multi-dataset)"><span>Schema for a pan-meta-genome graph with annotations (multi-dataset)</span></a></li><li><a class="tocitem" href="#Schema-for-a-pan-meta-genome-graph-(multi-dataset)"><span>Schema for a pan-meta-genome graph (multi-dataset)</span></a></li><li><a class="tocitem" href="#Schema-for-a-metagenome-graph-(single-dataset)"><span>Schema for a metagenome graph (single dataset)</span></a></li><li><a class="tocitem" href="#Example-Queries"><span>Example Queries</span></a></li><li><a class="tocitem" href="#Creating-a-blank-database"><span>Creating a blank database</span></a></li><li><a class="tocitem" href="#Interacting-with-a-Database"><span>Interacting with a Database</span></a></li><li><a class="tocitem" href="#Database"><span>Database</span></a></li><li><a class="tocitem" href="#Creating-probabilistic-assemblies"><span>Creating probabilistic assemblies</span></a></li><li><a class="tocitem" href="#Implementation-requirements"><span>Implementation requirements</span></a></li><li><a class="tocitem" href="#API"><span>API</span></a></li><li><a class="tocitem" href="#Learning-to-use-Cyper-CLI"><span>Learning to use Cyper CLI</span></a></li></ul></li><li><span class="tocitem">Documentation</span><ul><li><a class="tocitem" href="docstrings/">Docstrings</a></li></ul></li><li><span class="tocitem">Chapters</span><ul><li><a class="tocitem" href="1.using-graph-genomes-for-genome-optimization/">Using graph genomes for genome optimization</a></li><li><a class="tocitem" href="2.iterative-graph-polishing/">Iterative Graph polishing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Home</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Home</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/cjprybol/Mycelia.jl/blob/master/docs/src/index.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Mycelia"><a class="docs-heading-anchor" href="#Mycelia">Mycelia</a><a id="Mycelia-1"></a><a class="docs-heading-anchor-permalink" href="#Mycelia" title="Permalink"></a></h1><p>biological knowledge graphs for &quot;omics&quot; powered by <a href="https://github.com/JuliaGraphs">JuliaGraphs</a>, <a href="https://github.com/BioJulia">BioJulia</a>, and <a href="https://neo4j.com/">Neo4J</a></p><p><a href="http://research.cjp.garden/Mycelia/"><img src="https://img.shields.io/badge/docs-dev-blue.svg" alt="Dev"/></a> &lt;!– <a href="https://github.com/cjprybol/Mycelia.jl/pipelines"><img src="https://github.com/cjprybol/Mycelia.jl/badges/master/pipeline.svg" alt="Build Status"/></a> –&gt; &lt;!– <a href="https://github.com/cjprybol/Mycelia.jl/commits/master"><img src="https://github.com/cjprybol/Mycelia.jl/badges/master/coverage.svg" alt="Coverage"/></a> –&gt; &lt;!– <a href="https://github.com/SciML/ColPrac"><img src="https://img.shields.io/badge/ColPrac-Contributor&#39;s%20Guide-blueviolet" alt="ColPrac: Contributor&#39;s Guide on Collaborative Practices for Community Packages"/></a> –&gt;</p><h2 id="Install"><a class="docs-heading-anchor" href="#Install">Install</a><a id="Install-1"></a><a class="docs-heading-anchor-permalink" href="#Install" title="Permalink"></a></h2><pre><code class="language-julia">pkg&gt; add git@github.com:cjprybol/Mycelia.git</code></pre><h2 id="Related-Software:"><a class="docs-heading-anchor" href="#Related-Software:">Related Software:</a><a id="Related-Software:-1"></a><a class="docs-heading-anchor-permalink" href="#Related-Software:" title="Permalink"></a></h2><ul><li><a href="https://www.bioinformatics.nl/pangenomics/manual/">PanTools</a> (<a href="https://pubmed.ncbi.nlm.nih.gov/27587666/">paper</a>)</li><li><a href="https://github.com/BioJulia/GenomeGraphs.jl">BioJulia/GenomeGraphs.jl</a></li><li><a href="https://doi.org/10.1146/annurev-genom-120219-080406">Pangenome Graphs Review</a></li></ul><h2 id="Questions:"><a class="docs-heading-anchor" href="#Questions:">Questions:</a><a id="Questions:-1"></a><a class="docs-heading-anchor-permalink" href="#Questions:" title="Permalink"></a></h2><ul><li>what is the kmer decomposition spectra of raw reads @ 90, 95, 99, 99.9, 99.99% accuracy<ul><li>isolates?</li><li>communities?</li></ul></li><li>what are the sequence likelihoods when evaluating original reads vs their corrected versions<ul><li>before -&gt; after plots of alignment likelihood</li></ul></li></ul><h2 id="Behaviors"><a class="docs-heading-anchor" href="#Behaviors">Behaviors</a><a id="Behaviors-1"></a><a class="docs-heading-anchor-permalink" href="#Behaviors" title="Permalink"></a></h2><ul><li><p>Input(s):</p><ul><li>fasta(s)</li><li>fastq(s)</li></ul></li><li><p>Output(s):</p><ul><li>maximum likelihood assembly graph<ul><li>kmer frequencies</li></ul></li></ul></li><li><p>default assembly sizes = </p></li></ul><p>13  17  23  29  37  41  47  53  59</p><ul><li><p>Input(s):</p><ul><li>matrix of species frequency across samples</li></ul></li><li><p>Output(s):</p><ul><li>correlation matrix and strength of relationship</li></ul></li><li><p>Input(s):</p><ul><li>Genome query</li></ul></li><li><p>Output(s):</p><ul><li>closest entity</li><li>distance from that entity</li><li>context about what are the distances between organisms at different levels of the NCBI tree</li><li>lowest NCBI taxonomy classification at k 1-63</li><li>annotations contained within</li></ul></li><li><p>Input(s):</p></li><li><p>Output(s):</p></li></ul><h2 id="Schema-for-a-pan-meta-genome-graph-with-annotations-(multi-dataset)"><a class="docs-heading-anchor" href="#Schema-for-a-pan-meta-genome-graph-with-annotations-(multi-dataset)">Schema for a pan-meta-genome graph with annotations (multi-dataset)</a><a id="Schema-for-a-pan-meta-genome-graph-with-annotations-(multi-dataset)-1"></a><a class="docs-heading-anchor-permalink" href="#Schema-for-a-pan-meta-genome-graph-with-annotations-(multi-dataset)" title="Permalink"></a></h2><p><img src="pan-meta-genome-graph-with-annotations.svg" alt="pan-meta-genome-graph-with-annotations"/></p><pre><code class="language-none">//pan-meta-genome-graph-with-annotations
CREATE (dnamer:DNAmer {label: &quot;DNAmer&quot;})
CREATE (read:Read {label: &quot;Read&quot;})
CREATE (aamer:AAmer {label: &quot;AAmer&quot;})
CREATE (dataset:Dataset {label: &quot;Dataset&quot;})
CREATE (environment:Environment {label: &quot;Environment&quot;})
CREATE (genome:Genome {label: &quot;Genome&quot;})
CREATE (species:Species {label: &quot;Species&quot;})
CREATE (genus:Genus {label: &quot;Genus&quot;})
CREATE (family:Family {label: &quot;Family&quot;})
CREATE (order:Order {label: &quot;Order&quot;})
CREATE (subclass:Subclass {label: &quot;Subclass&quot;})
CREATE (class:Class {label: &quot;Class&quot;})
CREATE (subphylum:Subphylum {label: &quot;Subphylum&quot;})
CREATE (phylum:Phylum {label: &quot;Phylum&quot;})
CREATE (subkingdom:Subkingdom {label: &quot;Subkingdom&quot;})
CREATE (kingdom:Kingdom {label: &quot;Kingdom&quot;})
CREATE (clade:Clade {label: &quot;Clade&quot;})
CREATE (superkingdom:Superkingdom {label: &quot;Superkingdom&quot;})
CREATE (root:Root {label: &quot;Root&quot;})
CREATE (annotation:Annotation {label: &quot;Annotation&quot;})
CREATE (entity:Entity {label: &quot;Entity&quot;})
CREATE (pathway:Pathway {label: &quot;Pathway&quot;})
CREATE (toxin:Toxin {label:&quot;Toxin&quot;})
CREATE (virulencefactor:VirulenceFactor {label: &quot;VirulenceFactor&quot;})
CREATE (antibioticresistance:AntibioticResistance {label: &quot;AntibioticResistance&quot;})
CREATE (integrase:Integrase {label: &quot;Integrase&quot;})
CREATE (host:Entity {label: &quot;Host&quot;})

CREATE (entity)-[entity_host:HOST_OF]-&gt;(host)
CREATE (annotation)-[annotation_integrase:ISA]-&gt;(integrase)
CREATE (annotation)-[annotation_antibioticresistance:CONFERS]-&gt;(antibioticresistance)
CREATE (annotation)-[annotation_virulencefator:ISA]-&gt;(virulencefactor)
CREATE (annotation)-[annotation_toxin:ISA]-&gt;(toxin)
CREATE (host)-[host_species:ISA]-&gt;(species)
CREATE (host)-[host_genome:GENOME_OF]-&gt;(genome)
CREATE (dnamer)-[dnamer_annotation:CONTAINED_IN]-&gt;(annotation)
CREATE (aamer)-[aamer_annotation:CONTAINED_IN]-&gt;(annotation)
CREATE (genome)-[genome_annotation:CONTAINED_IN]-&gt;(annotation)
CREATE (superkingdom)-[superkingdom_root:PARENT]-&gt;(root)
CREATE (clade)-[clade_superkingdom:PARENT]-&gt;(superkingdom)
CREATE (kingdom)-[kingdom_clade:PARENT]-&gt;(clade)
CREATE (subkingdom)-[subkingdom_kingdom:PARENT]-&gt;(kingdom)
CREATE (phylum)-[phylum_subkingdom:PARENT]-&gt;(subkingdom)
CREATE (subphylum)-[subphylum_phylum:PARENT]-&gt;(phylum)
CREATE (class)-[class_subphylum:PARENT]-&gt;(subphylum)
CREATE (subclass)-[subclass_class:PARENT]-&gt;(class)
CREATE (order)-[order_subclass:PARENT]-&gt;(subclass)
CREATE (family)-[family_order:PARENT]-&gt;(order)
CREATE (genus)-[genus_family:PARENT]-&gt;(family)
CREATE (species)-[species_genus:PARENT]-&gt;(genus)
CREATE (entity)-[entity_species:ISA]-&gt;(species)
CREATE (dnamer)-[dnamer_genome:CONTAINED_IN]-&gt;(genome)
CREATE (dataset)-[dataset_environment:SOURCED_FROM]-&gt;(environment)
CREATE (read)-[read_dataset:CONTAINED_IN]-&gt;(dataset)
CREATE (aamer)-[aamer_read:CONTAINED_IN]-&gt;(read)
CREATE (dnamer)-[dnamer_read:CONTAINED_IN]-&gt;(read)
CREATE (dnamer)-[dnamer_aamer:TRANSLATES_TO]-&gt;(aamer)
CREATE (aamer)-[aamer_genome:CONTAINED_IN]-&gt;(genome)
CREATE (genome)-[genome_entity:GENOME_OF]-&gt;(entity)
CREATE (annotation)-[annotation_pathway:CONTAINED_IN]-&gt;(pathway)


RETURN dnamer,
read,
aamer,
dataset,
environment,
genome,
species,
genus,
family,
order,
subclass,
class,
subphylum,
phylum,
subkingdom,
kingdom,
clade,
superkingdom,
root,
annotation,
entity,
pathway,
annotation_pathway,
toxin,
virulencefactor,
antibioticresistance,
integrase,
host,
entity_host,
annotation_integrase,
annotation_antibioticresistance,
annotation_virulencefator,
annotation_toxin,
host_species,
host_genome
</code></pre><h2 id="Schema-for-a-pan-meta-genome-graph-(multi-dataset)"><a class="docs-heading-anchor" href="#Schema-for-a-pan-meta-genome-graph-(multi-dataset)">Schema for a pan-meta-genome graph (multi-dataset)</a><a id="Schema-for-a-pan-meta-genome-graph-(multi-dataset)-1"></a><a class="docs-heading-anchor-permalink" href="#Schema-for-a-pan-meta-genome-graph-(multi-dataset)" title="Permalink"></a></h2><p><img src="pan-meta-genome-graph.svg" alt="pan-meta-genome-graph"/></p><pre><code class="language-none">//pan-meta-genome-graph
CREATE (dnamer:DNAmer {label: &quot;DNAmer&quot;})
CREATE (read:Read {label: &quot;Read&quot;})
CREATE (aamer:AAmer {label: &quot;AAmer&quot;})
CREATE (dataset:Dataset {label: &quot;Dataset&quot;})
CREATE (environment:Environment {label: &quot;Environment&quot;})
CREATE (genome:Genome {label: &quot;Genome&quot;})
CREATE (species:Species {label: &quot;Species&quot;})
CREATE (genus:Genus {label: &quot;Genus&quot;})
CREATE (family:Family {label: &quot;Family&quot;})
CREATE (order:Order {label: &quot;Order&quot;})
CREATE (subclass:Subclass {label: &quot;Subclass&quot;})
CREATE (class:Class {label: &quot;Class&quot;})
CREATE (subphylum:Subphylum {label: &quot;Subphylum&quot;})
CREATE (phylum:Phylum {label: &quot;Phylum&quot;})
CREATE (subkingdom:Subkingdom {label: &quot;Subkingdom&quot;})
CREATE (kingdom:Kingdom {label: &quot;Kingdom&quot;})
CREATE (clade:Clade {label: &quot;Clade&quot;})
CREATE (superkingdom:Superkingdom {label: &quot;Superkingdom&quot;})
CREATE (root:Root {label: &quot;Root&quot;})
CREATE (annotation:Annotation {label: &quot;Annotation&quot;})
CREATE (entity:Entity {label: &quot;Entity&quot;})
CREATE (pathway:Pathway {label: &quot;Pathway&quot;})

CREATE (dnamer)-[dnamer_annotation:CONTAINED_IN]-&gt;(annotation)
CREATE (aamer)-[aamer_annotation:CONTAINED_IN]-&gt;(annotation)
CREATE (genome)-[genome_annotation:CONTAINED_IN]-&gt;(annotation)
CREATE (superkingdom)-[superkingdom_root:PARENT]-&gt;(root)
CREATE (clade)-[clade_superkingdom:PARENT]-&gt;(superkingdom)
CREATE (kingdom)-[kingdom_clade:PARENT]-&gt;(clade)
CREATE (subkingdom)-[subkingdom_kingdom:PARENT]-&gt;(kingdom)
CREATE (phylum)-[phylum_subkingdom:PARENT]-&gt;(subkingdom)
CREATE (subphylum)-[subphylum_phylum:PARENT]-&gt;(phylum)
CREATE (class)-[class_subphylum:PARENT]-&gt;(subphylum)
CREATE (subclass)-[subclass_class:PARENT]-&gt;(class)
CREATE (order)-[order_subclass:PARENT]-&gt;(subclass)
CREATE (family)-[family_order:PARENT]-&gt;(order)
CREATE (genus)-[genus_family:PARENT]-&gt;(family)
CREATE (species)-[species_genus:PARENT]-&gt;(genus)
CREATE (entity)-[entity_species:ISA]-&gt;(species)
CREATE (dnamer)-[dnamer_genome:CONTAINED_IN]-&gt;(genome)
CREATE (dataset)-[dataset_environment:SOURCED_FROM]-&gt;(environment)
CREATE (read)-[read_dataset:CONTAINED_IN]-&gt;(dataset)
CREATE (aamer)-[aamer_read:CONTAINED_IN]-&gt;(read)
CREATE (dnamer)-[dnamer_read:CONTAINED_IN]-&gt;(read)
CREATE (dnamer)-[dnamer_aamer:TRANSLATES_TO]-&gt;(aamer)
CREATE (aamer)-[aamer_genome:CONTAINED_IN]-&gt;(genome)
CREATE (genome)-[genome_entity:GENOME_OF]-&gt;(entity)
CREATE (annotation)-[annotation_pathway:CONTAINED_IN]-&gt;(pathway)


RETURN dnamer,
read,
aamer,
dataset,
environment,
genome,
species,
genus,
family,
order,
subclass,
class,
subphylum,
phylum,
subkingdom,
kingdom,
clade,
superkingdom,
root,
annotation,
entity,
pathway,
annotation_pathway</code></pre><h2 id="Schema-for-a-metagenome-graph-(single-dataset)"><a class="docs-heading-anchor" href="#Schema-for-a-metagenome-graph-(single-dataset)">Schema for a metagenome graph (single dataset)</a><a id="Schema-for-a-metagenome-graph-(single-dataset)-1"></a><a class="docs-heading-anchor-permalink" href="#Schema-for-a-metagenome-graph-(single-dataset)" title="Permalink"></a></h2><p><img src="meta-genome-graph.svg" alt="meta-genome-graph"/></p><pre><code class="language-none">//meta-genome-graph
CREATE (dnamer:DNAmer {label: &quot;DNAmer&quot;})
CREATE (aamer:AAmer {label: &quot;AAmer&quot;})
CREATE (read:Read {label: &quot;Read&quot;})
CREATE (genome:Genome {label: &quot;Genome&quot;})
CREATE (annotation:Annotation {label: &quot;Annotation&quot;})
CREATE (pathway:Pathway {label: &quot;Pathway&quot;})

CREATE (dnamer)-[dnamer_annotation:CONTAINED_IN]-&gt;(annotation)
CREATE (aamer)-[aamer_annotation:CONTAINED_IN]-&gt;(annotation)
CREATE (genome)-[genome_annotation:CONTAINED_IN]-&gt;(annotation)
CREATE (annotation)-[annotation_pathway:CONTAINED_IN]-&gt;(pathway)
CREATE (dnamer)-[dnamer_genome:CONTAINED_IN]-&gt;(genome)
CREATE (aamer)-[aamer_read:CONTAINED_IN]-&gt;(read)
CREATE (dnamer)-[dnamer_read:CONTAINED_IN]-&gt;(read)
CREATE (dnamer)-[dnamer_aamer:TRANSLATES_TO]-&gt;(aamer)
CREATE (aamer)-[aamer_genome:CONTAINED_IN]-&gt;(genome)

RETURN dnamer,
read,
aamer,
genome,
annotation,
pathway,
annotation_pathway</code></pre><h2 id="Example-Queries"><a class="docs-heading-anchor" href="#Example-Queries">Example Queries</a><a id="Example-Queries-1"></a><a class="docs-heading-anchor-permalink" href="#Example-Queries" title="Permalink"></a></h2><p>English</p><pre><code class="language-none">CYPHER</code></pre><ul><li><p>Give me the genome of entity x</p></li><li><p>Give me the pangenome of species x</p></li><li><p>Give me the pangenome of order x</p></li><li><p>Give me all kmers in species x</p></li><li><p>Finding spacers:</p></li><li><p>Give me all kmers in species x matching pattern y</p></li><li><p>Give me all datasets containing species x</p></li><li><p>Give me all datasets containing pathway x</p></li><li><p>Give me all reads supporting path/variant x</p></li></ul><h2 id="Creating-a-blank-database"><a class="docs-heading-anchor" href="#Creating-a-blank-database">Creating a blank database</a><a id="Creating-a-blank-database-1"></a><a class="docs-heading-anchor-permalink" href="#Creating-a-blank-database" title="Permalink"></a></h2><p>https://neo4j.com/developer/neo4j-desktop/#desktop-create-project</p><h2 id="Interacting-with-a-Database"><a class="docs-heading-anchor" href="#Interacting-with-a-Database">Interacting with a Database</a><a id="Interacting-with-a-Database-1"></a><a class="docs-heading-anchor-permalink" href="#Interacting-with-a-Database" title="Permalink"></a></h2><ul><li>username: neo4j</li><li>password: password<ul><li>I set this and yours may be different!</li></ul></li></ul><p>https://neo4j.com/developer/manage-multiple-databases/</p><p>https://stackoverflow.com/a/29658062 https://neo4j.com/docs/operations-manual/current/tools/cypher-shell/#cypher-shell-syntax https://contentaudience.com/guides/neo4j-cli-cypher-shell/ https://neo4j.com/developer/kb/how-do-i-authenticate-with-cypher-shell-without-specifying-the-username-and-password-on-the-command-line/</p><p><a href="https://neo4j.com/developer/guide-import-csv/">Bulk import</a></p><h2 id="Database"><a class="docs-heading-anchor" href="#Database">Database</a><a id="Database-1"></a><a class="docs-heading-anchor-permalink" href="#Database" title="Permalink"></a></h2><ul><li>Basic Alphabets<ul><li>Canonical Nucleic Acids</li><li>Canonical Amino Acids</li></ul></li><li>Types of paths<ul><li>DNAmers<ul><li>DNAmers are canonical dna sequences of k-length</li><li>we track all odd primes between 1 and 31 for DNA mers</li><li>we track DNAmers 3, 9, 15, and 21 for interoperability between AAmers and the DNAmers they translate from</li><li>kmers store both their full sequences AND their sequences as paths through shorter kmers</li></ul></li><li>AAmers<ul><li>we track all odd primes between 1 and 5 for AA mers</li><li>kmers store both their full sequences AND their sequences as paths through shorter kmers</li></ul></li><li>Genomes<ul><li>genomes are paths through DNA kmers</li></ul></li><li>annotations<ul><li>annotations are paths through DNA kmers and possibly also through AA kmers</li></ul></li></ul></li></ul><h2 id="Creating-probabilistic-assemblies"><a class="docs-heading-anchor" href="#Creating-probabilistic-assemblies">Creating probabilistic assemblies</a><a id="Creating-probabilistic-assemblies-1"></a><a class="docs-heading-anchor-permalink" href="#Creating-probabilistic-assemblies" title="Permalink"></a></h2><ol><li>Build weighted <a href="https://en.wikipedia.org/wiki/De_Bruijn_graph">de-bruijn graphs</a> with observed data</li><li>Use the weighted de-bruijn graph as a <a href="https://en.wikipedia.org/wiki/Hidden_Markov_model">hidden markov model</a> to <a href="https://en.wikipedia.org/wiki/Viterbi_algorithm">error correct observations</a></li><li>Use the error-corrected observations to build a new, more accurate weighted de-bruijn graph</li><li>Repeat 1-3 until convergence</li><li>Return maximum likelihood assembly</li></ol><p>Consider for graph cleaning:</p><ul><li>[x] Solve most likely path (slow but robust) vs</li><li>[ ] resample most likely paths (fast, less theoretically sound)<ul><li>fairly certain that this is what the Flye assembler does</li></ul></li></ul><h2 id="Implementation-requirements"><a class="docs-heading-anchor" href="#Implementation-requirements">Implementation requirements</a><a id="Implementation-requirements-1"></a><a class="docs-heading-anchor-permalink" href="#Implementation-requirements" title="Permalink"></a></h2><ul><li><p>kmer spectra -&gt; signal detection</p></li><li><p>kmer size with signal -&gt; initial assembly graph</p><ul><li>bandage visualization</li></ul></li><li><p>initial assembly graph + reads -&gt; iterative assembly</p><ul><li>bandage visualization</li></ul></li><li><p>final assembly</p><ul><li>bandage visualization</li></ul></li><li><p>annotation</p><ul><li>find all start codons</li><li>find all stop codons</li><li>report in fasta format all coding regions</li><li>try and verify against prodigal to make sure that we aren’t missing any</li></ul></li><li><p>graph -&gt; path -&gt; sequence -&gt; amino acid sequence</p></li></ul><h2 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h2><ul><li>Need to allow users to add individual datasets to database</li><li>does a sequence exist?<ul><li>does a sequence exist allowing n mismatches</li><li>relative likelihood and frequency of most similar paths<ul><li>this is just the viterbi traversal</li></ul></li></ul></li><li>reads supporting traversal, and the datasets they came from</li><li>connected components (species) containing path</li><li>phylogenetic hierarchy of a component<ul><li>nodes &lt;-&gt; kmers</li></ul></li><li>which datasets are a given kmer present in?</li><li>export and import with GFA</li><li>condense kmer graphs to simplified graphs</li><li>visualization <ul><li>with Bandage (works automatically with GFA)</li><li>with Neo4J visualization library</li></ul></li></ul><h2 id="Learning-to-use-Cyper-CLI"><a class="docs-heading-anchor" href="#Learning-to-use-Cyper-CLI">Learning to use Cyper CLI</a><a id="Learning-to-use-Cyper-CLI-1"></a><a class="docs-heading-anchor-permalink" href="#Learning-to-use-Cyper-CLI" title="Permalink"></a></h2><pre><code class="language-none">usage: cypher-shell [-h] [-a ADDRESS] [-u USERNAME] [-p PASSWORD] [--encryption {true,false,default}] [-d DATABASE]
                    [--format {auto,verbose,plain}] [-P PARAM] [--debug] [--non-interactive] [--sample-rows SAMPLE-ROWS]
                    [--wrap {true,false}] [-v] [--driver-version] [-f FILE] [--fail-fast | --fail-at-end] [cypher]

A command line shell where you can execute Cypher against an  instance  of  Neo4j.  By default the shell is interactive but you can use it
for scripting by passing cypher directly on the command line or by piping a file with cypher statements (requires Powershell on Windows).

example of piping a file:
  cat some-cypher.txt | cypher-shell

positional arguments:
  cypher                 an optional string of cypher to execute and then exit

optional arguments:
  -h, --help             show this help message and exit
  --fail-fast            exit and report failure on first error when reading from file (this is the default behavior)
  --fail-at-end          exit and report failures at end of input when reading from file
  --format {auto,verbose,plain}
                         desired output format, verbose displays results  in  tabular  format  and  prints statistics, plain displays data
                         with minimal formatting (default: auto)
  -P PARAM, --param PARAM
                         Add a parameter to this session. Example: `-P &quot;number =&gt; 3&quot;`. This argument can be specified multiple times.
  --debug                print additional debug information (default: false)
  --non-interactive      force non-interactive mode, only useful if auto-detection fails (like on Windows) (default: false)
  --sample-rows SAMPLE-ROWS
                         number of rows sampled to compute table widths (only for format=VERBOSE) (default: 1000)
  --wrap {true,false}    wrap table column values if column is too narrow (only for format=VERBOSE) (default: true)
  -v, --version          print version of cypher-shell and exit (default: false)
  --driver-version       print version of the Neo4j Driver used and exit (default: false)
  -f FILE, --file FILE   Pass a file with cypher statements to be executed.  After  the statements have been executed cypher-shell will be
                         shutdown

connection arguments:
  -a ADDRESS, --address ADDRESS
                         address and port to connect to (default: neo4j://localhost:7687)
  -u USERNAME, --username USERNAME
                         username to connect as. Can also be specified using environment variable NEO4J_USERNAME (default: )
  -p PASSWORD, --password PASSWORD
                         password to connect with. Can also be specified using environment variable NEO4J_PASSWORD (default: )
  --encryption {true,false,default}
                         whether the connection to Neo4j should  be  encrypted.  This  must  be  consistent with Neo4j&#39;s configuration. If
                         choosing &#39;default&#39; the encryption setting is  deduced  from  the  specified  address. For example the &#39;neo4j+ssc&#39;
                         protocol would use encryption. (default: default)
  -d DATABASE, --database DATABASE
                         database to connect to. Can also be specified using environment variable NEO4J_DATABASE (default: )</code></pre><p>List databases</p><pre><code class="language-none"># verbose outputs mdtable with pretty printing
cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database system --format verbose &#39;show databases&#39;
# plain outputs a CSV
cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database system --format plain &#39;show databases&#39;
# auto uses verbose
cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database system --format auto &#39;show databases&#39;</code></pre><p>Creating a database for a specific dataset</p><pre><code class="language-none">cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database system --format auto &#39;create database test&#39;</code></pre><p>Show schema</p><pre><code class="language-none">cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto &#39;CALL db.schema.visualization()&#39;</code></pre><p>Return all nodes</p><pre><code class="language-none">cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto &#39;MATCH (n) RETURN n&#39;</code></pre><p>Delete entire graph</p><pre><code class="language-none">cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto &#39;MATCH (n) DETACH DELETE n&#39;</code></pre><p>Add kmers and connections in one go</p><pre><code class="language-none">cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto &#39;MERGE (d1:DNAmer {sequence: &quot;A&quot;})-[d1d2:CONNECTION {orientations: [1,1]}]-&gt;(d2:DNAmer {sequence: &quot;T&quot;}) RETURN *&#39;</code></pre><p>Add kmers and connections in separate steps</p><pre><code class="language-none">cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto &#39;MERGE (:DNAmer {sequence: &quot;A&quot;})&#39;
cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto &#39;MERGE (:DNAmer {sequence: &quot;T&quot;})&#39;
cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto &#39;MERGE (d1:DNAmer {sequence: &quot;A&quot;})-[d1d2:CONNECTION {orientations: [1,1]}]-&gt;(d2:DNAmer {sequence: &quot;T&quot;}) RETURN *&#39;</code></pre><p>Find the import folder on </p><pre><code class="language-none"># https://neo4j.com/docs/operations-manual/current/configuration/file-locations/
/Users/cameronprybol/Library/Application Support/Neo4j Desktop/Application/relate-data/dbmss/dbms-8ab8baac-5dea-4137-bb24-e0b426447940/import</code></pre><p>find all nmers</p><pre><code class="language-none">cypher-shell --address neo4j://localhost:7687 --username neo4j --password password --database test --format auto &#39;MATCH (d:DNAmer) WHERE size(d.sequence) = 1 return d&#39;</code></pre><p>create import uniqueness constraints</p><pre><code class="language-none">CREATE CONSTRAINT ON (d:Dnamer) ASSERT d.sequence IS UNIQUE;
CREATE CONSTRAINT ON (f:Fasta) ASSERT f.identifier IS UNIQUE;
CREATE CONSTRAINT ON (f:Fastq) ASSERT f.identifier IS UNIQUE;</code></pre><p>data import</p><pre><code class="language-none">USING PERIODIC COMMIT
LOAD CSV WITH HEADERS FROM
&#39;file:///path/to/file&#39; AS line
with line

CREATE (n:Node {id: line.`id`})</code></pre><p>data conversions for loading CSV</p><pre><code class="language-none">TOINT
TOFLOAT
https://neo4j.com/docs/cypher-manual/current/functions/</code></pre><p>merge nodes to avoid creating again</p><p>!important!</p><p>When merging new information into nodes, need to first match on unique key to get all existing fields. Merging matches on all fields, and merging on partial matches will duplicate nodes which we don&#39;t want</p></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="docstrings/">Docstrings »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Wednesday 15 December 2021 00:44">Wednesday 15 December 2021</span>. Using Julia version 1.6.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
