<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>K primes pattern · Mycelia</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://cjprybol.github.io/Mycelia/4.Reference/k-primes-pattern.ipynb/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">Mycelia</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><input class="collapse-toggle" id="menuitem-2" type="checkbox"/><label class="tocitem" for="menuitem-2"><span class="docs-label">1.Tutorial</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../1.Tutorial/0.constructing-graphs-from-observed-sequences.ipynb/">Constructing Graphs from Observed Sequences</a></li><li><a class="tocitem" href="../../1.Tutorial/1.constructing-graphs-from-reference-sequences.ipynb/">Constructing Graphs From Reference Sequences</a></li><li><a class="tocitem" href="../../1.Tutorial/25.converting-graph-formats.ipynb/">Converting Graph Formats</a></li><li><a class="tocitem" href="../../1.Tutorial/3.merging-graphs.ipynb/">Merging Graphs</a></li><li><a class="tocitem" href="../../1.Tutorial/5.filtering-graphs.ipynb/">Filtering Graphs</a></li><li><a class="tocitem" href="../../1.Tutorial/6.annotating-graphs.ipynb/">Annotating Graphs</a></li><li><a class="tocitem" href="../../1.Tutorial/7.reducing-graphs.ipynb/">Reducing Graphs</a></li><li><a class="tocitem" href="../../1.Tutorial/8.finding-consensus-sequences.ipynb/">Finding consensus sequences</a></li><li><a class="tocitem" href="../../1.Tutorial/9.visualizing-graphs.ipynb/">Visualizing Graphs</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox"/><label class="tocitem" for="menuitem-3"><span class="docs-label">2.Applications</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../2.Applications/1.hybrid-metagenomic-assembly.ipynb/">Hybrid Metagenomic Assembly</a></li><li><a class="tocitem" href="../../2.Applications/2.minimal-genome-inference.ipynb/">Minimal Genome Inference</a></li><li><a class="tocitem" href="../../2.Applications/3.microbiome-analysis.ipynb/">Microbiome Analysis</a></li><li><a class="tocitem" href="../../2.Applications/4.microbiome-modification.ipynb/">Microbiome Modification</a></li><li><a class="tocitem" href="../../2.Applications/5.clustering.ipynb/">Clustering</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-4" type="checkbox"/><label class="tocitem" for="menuitem-4"><span class="docs-label">3.Theory</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../3.Theory/00.limitations-of-linear-reference-omes.ipynb/">Limitations of Linear Reference (Gen/Transcript/Prote)omes</a></li><li><a class="tocitem" href="../../3.Theory/005.representing-sequence-with-kmers.ipynb/">Representing Sequences with Kmers</a></li><li><a class="tocitem" href="../../3.Theory/01.representating-omes-with-graphs.ipynb/">Representing (Gen/Transcript/Prote)omes with Graphs</a></li><li><a class="tocitem" href="../../3.Theory/025.representing-omics-graphs-with-metagraphs.ipynb/">Representing Multiomics Graphs with MetaGraphs</a></li><li><a class="tocitem" href="../../3.Theory/03.optimizing-graphs.ipynb/">Using graph genomes for genome optimization</a></li><li><a class="tocitem" href="../../3.Theory/04.graph-simplification.ipynb/">Graph Simplification</a></li><li><a class="tocitem" href="../../3.Theory/095.clustering-and-classification.ipynb/">Clustering and Classification</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-5" type="checkbox" checked/><label class="tocitem" for="menuitem-5"><span class="docs-label">4.Reference</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../development-notes.ipynb/">Development Notes</a></li><li><a class="tocitem" href="../docstrings.ipynb/">Docstrings</a></li><li class="is-active"><a class="tocitem" href>K primes pattern</a></li><li><a class="tocitem" href="../neo4j-notes.ipynb/">Neo4J Notes</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">4.Reference</a></li><li class="is-active"><a href>K primes pattern</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>K primes pattern</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/cjprybol/Mycelia/blob/master/docs/src/4.Reference/k-primes-pattern.ipynb.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="K-primes-pattern"><a class="docs-heading-anchor" href="#K-primes-pattern">K primes pattern</a><a id="K-primes-pattern-1"></a><a class="docs-heading-anchor-permalink" href="#K-primes-pattern" title="Permalink"></a></h1><p>I was looking for a way to iterate through prime number k&#39;s. Odd-length kmers are desireable because they cannot be reverse complements of themselves [citeme]. Primes are desireable because they cannot be perfect repeats. An ideal iteration pattern would avoid including both twins in <a href="https://en.wikipedia.org/wiki/Twin_prime">twin prime</a> pairs (I have no specific preference about whether the chosen is the larger or the smaller). Re-evaluating data with both twins in a twin pair, where the windows are so similarly sized, seems like it would be for minimal gain. With the same principle of trying to avoid reprocessing data for minimal gain, an ideal iteration would steadily increase in distance between the current k-length and the next. I stumbled upon this pattern, which may be a useful way to limit the analysis space.</p><pre><code class="language-julia hljs">julia&gt; step = 2
2

julia&gt; current_prime = 11
11

julia&gt; while Primes.isprime(current_prime)
           current_prime = current_prime + step
           step += 2
           @show current_prime
       end
current_prime = 13
current_prime = 17
current_prime = 23
current_prime = 31
current_prime = 41
current_prime = 53
current_prime = 67
current_prime = 83
current_prime = 101
current_prime = 121</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../docstrings.ipynb/">« Docstrings</a><a class="docs-footer-nextpage" href="../neo4j-notes.ipynb/">Neo4J Notes »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.10 on <span class="colophon-date" title="Thursday 23 December 2021 02:49">Thursday 23 December 2021</span>. Using Julia version 1.6.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
