<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Development Notes · Mycelia</title><script data-outdated-warner src="../../assets/warner.js"></script><link rel="canonical" href="https://cjprybol.github.io/Mycelia/4.Reference/development-notes.ipynb/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">Mycelia</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><input class="collapse-toggle" id="menuitem-2" type="checkbox"/><label class="tocitem" for="menuitem-2"><span class="docs-label">1.Tutorial</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../1.Tutorial/0.constructing-graphs-from-observed-sequences.ipynb/">Constructing Graphs from Observed Sequences</a></li><li><a class="tocitem" href="../../1.Tutorial/1.constructing-graphs-from-reference-sequences.ipynb/">Constructing Graphs From Reference Sequences</a></li><li><a class="tocitem" href="../../1.Tutorial/25.converting-graph-formats.ipynb/">Converting Graph Formats</a></li><li><a class="tocitem" href="../../1.Tutorial/3.merging-graphs.ipynb/">Merging Graphs</a></li><li><a class="tocitem" href="../../1.Tutorial/5.filtering-graphs.ipynb/">Filtering Graphs</a></li><li><a class="tocitem" href="../../1.Tutorial/6.annotating-graphs.ipynb/">Annotating Graphs</a></li><li><a class="tocitem" href="../../1.Tutorial/7.reducing-graphs.ipynb/">Reducing Graphs</a></li><li><a class="tocitem" href="../../1.Tutorial/8.finding-consensus-sequences.ipynb/">Finding consensus sequences</a></li><li><a class="tocitem" href="../../1.Tutorial/9.visualizing-graphs.ipynb/">Visualizing Graphs</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox"/><label class="tocitem" for="menuitem-3"><span class="docs-label">2.Applications</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../2.Applications/1.hybrid-metagenomic-assembly.ipynb/">Hybrid Metagenomic Assembly</a></li><li><a class="tocitem" href="../../2.Applications/2.minimal-prokaryotic-genome.ipynb/">Minimal Prokaryotic Genome</a></li><li><a class="tocitem" href="../../2.Applications/3.minimal-eukaryotic-chromosome.ipynb/">Minimal Eukaryotic Genome</a></li><li><a class="tocitem" href="../../2.Applications/4.microbiome-modification.ipynb/">Microbiome Modification</a></li><li><a class="tocitem" href="../../2.Applications/5.identifying-related-clusters.ipynb/">Identifying Related Clusters</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-4" type="checkbox"/><label class="tocitem" for="menuitem-4"><span class="docs-label">3.Discussion</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../3.Discussion/00.limitations-of-linear-reference-omes.ipynb/">Limitations of Linear Reference (Gen/Transcript/Prote)omes</a></li><li><a class="tocitem" href="../../3.Discussion/005.representing-sequence-with-kmers.ipynb/">Representing Sequences with Kmers</a></li><li><a class="tocitem" href="../../3.Discussion/01.representating-omes-with-graphs.ipynb/">Representing (Gen/Transcript/Prote)omes with Graphs</a></li><li><a class="tocitem" href="../../3.Discussion/025.representing-omics-graphs-with-metagraphs.ipynb/">Representing Multiomics Graphs with MetaGraphs</a></li><li><a class="tocitem" href="../../3.Discussion/03.optimizing-genome-graphs.ipynb/">Using graph genomes for genome optimization</a></li><li><a class="tocitem" href="../../3.Discussion/04.graph-simplification.ipynb/">Graph Simplification</a></li><li><a class="tocitem" href="../../3.Discussion/045.minimal-essential-genome.ipynb/">Minimal Essential Genome</a></li><li><a class="tocitem" href="../../3.Discussion/05.generalized-genome-optimization.ipynb/">Generalized Genome Optimization</a></li><li><a class="tocitem" href="../../3.Discussion/09.optimizing-microbiomes.ipynb/">Optimizing Microbiomes</a></li><li><a class="tocitem" href="../../3.Discussion/095.classifying-and-clustering-sequences-with-graphs.ipynb/">Classifying and Clustering Sequences with Graphs</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-5" type="checkbox" checked/><label class="tocitem" for="menuitem-5"><span class="docs-label">4.Reference</span><i class="docs-chevron"></i></label><ul class="collapsed"><li class="is-active"><a class="tocitem" href>Development Notes</a><ul class="internal"><li><a class="tocitem" href="#API"><span>API</span></a></li><li><a class="tocitem" href="#Implementation-requirements"><span>Implementation requirements</span></a></li><li><a class="tocitem" href="#Creating-probabilistic-assemblies"><span>Creating probabilistic assemblies</span></a></li></ul></li><li><a class="tocitem" href="../docstrings.ipynb/">Docstrings</a></li><li><a class="tocitem" href="../k-primes-pattern.ipynb/">K primes pattern</a></li><li><a class="tocitem" href="../neo4j-notes.ipynb/">Neo4J Notes</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">4.Reference</a></li><li class="is-active"><a href>Development Notes</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Development Notes</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/cjprybol/Mycelia/blob/master/docs/src/4.Reference/development-notes.ipynb.md#L" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Development-Notes"><a class="docs-heading-anchor" href="#Development-Notes">Development Notes</a><a id="Development-Notes-1"></a><a class="docs-heading-anchor-permalink" href="#Development-Notes" title="Permalink"></a></h1><h2 id="API"><a class="docs-heading-anchor" href="#API">API</a><a id="API-1"></a><a class="docs-heading-anchor-permalink" href="#API" title="Permalink"></a></h2><ul><li>Need to allow users to add individual datasets to database</li><li>does a sequence exist?<ul><li>does a sequence exist allowing n mismatches</li><li>relative likelihood and frequency of most similar paths<ul><li>this is just the viterbi traversal</li></ul></li></ul></li><li>reads supporting traversal, and the datasets they came from</li><li>connected components (species) containing path</li><li>phylogenetic hierarchy of a component<ul><li>nodes &lt;-&gt; kmers</li></ul></li><li>which datasets are a given kmer present in?</li><li>export and import with GFA</li><li>condense kmer graphs to simplified graphs</li><li>visualization <ul><li>with Bandage (works automatically with GFA)</li><li>with Neo4J visualization library</li></ul></li></ul><h2 id="Implementation-requirements"><a class="docs-heading-anchor" href="#Implementation-requirements">Implementation requirements</a><a id="Implementation-requirements-1"></a><a class="docs-heading-anchor-permalink" href="#Implementation-requirements" title="Permalink"></a></h2><ul><li><p>kmer spectra -&gt; signal detection</p></li><li><p>kmer size with signal -&gt; initial assembly graph</p><ul><li>bandage visualization</li></ul></li><li><p>initial assembly graph + reads -&gt; iterative assembly</p><ul><li>bandage visualization</li></ul></li><li><p>final assembly</p><ul><li>bandage visualization</li></ul></li><li><p>annotation</p><ul><li>find all start codons</li><li>find all stop codons</li><li>report in fasta format all coding regions</li><li>try and verify against prodigal to make sure that we aren’t missing any</li></ul></li><li><p>graph -&gt; path -&gt; sequence -&gt; amino acid sequence</p></li></ul><h2 id="Creating-probabilistic-assemblies"><a class="docs-heading-anchor" href="#Creating-probabilistic-assemblies">Creating probabilistic assemblies</a><a id="Creating-probabilistic-assemblies-1"></a><a class="docs-heading-anchor-permalink" href="#Creating-probabilistic-assemblies" title="Permalink"></a></h2><ol><li>Build weighted <a href="https://en.wikipedia.org/wiki/De_Bruijn_graph">de-bruijn graphs</a> with observed data</li><li>Use the weighted de-bruijn graph as a <a href="https://en.wikipedia.org/wiki/Hidden_Markov_model">hidden markov model</a> to <a href="https://en.wikipedia.org/wiki/Viterbi_algorithm">error correct observations</a></li><li>Use the error-corrected observations to build a new, more accurate weighted de-bruijn graph</li><li>Repeat 1-3 until convergence</li><li>Return maximum likelihood assembly</li></ol><p>Consider for graph cleaning:</p><ul><li>[x] Solve most likely path (slow but robust) vs</li><li>[ ] resample most likely paths (fast, less theoretically sound)<ul><li>fairly certain that this is what the Flye assembler does</li></ul></li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../../3.Discussion/095.classifying-and-clustering-sequences-with-graphs.ipynb/">« Classifying and Clustering Sequences with Graphs</a><a class="docs-footer-nextpage" href="../docstrings.ipynb/">Docstrings »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.10 on <span class="colophon-date" title="Monday 20 December 2021 18:08">Monday 20 December 2021</span>. Using Julia version 1.6.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
